File: ros/ws01/first_ws/src/ctrl_turtle_pkg/src/VelocityWindow.cpp
//
// Created by zcb on 2021/4/3.
//

#include "VelocityWindow.h"
#include <geometry_msgs/Twist.h>


VelocityWindow::VelocityWindow(ros::NodeHandle* node,QWidget* parent): QWidget(parent){
    this->setWindowTitle("小乌龟控制");

    // layout
    this->formlayout = new QFormLayout(this);

    this->editLinear = new QLineEdit("0.0");
    this->editAngular =new QLineEdit("0.0");
    this->formlayout->addRow("线速度", editLinear);
    this->formlayout->addRow("角速度", editAngular);

    this->sendBtn = new QPushButton("发送");
    this->formlayout->addRow(this->sendBtn);

    // event:sendBtn
    QPushButton::connect(this->sendBtn, &QPushButton::clicked,this, &VelocityWindow::sendMsgEvt);

    std::string topicName = "/turtle1/cmd_vel";
    this->publisher = node->advertise<geometry_msgs::Twist>(topicName, 1000);
}

VelocityWindow::~VelocityWindow() {

}

void VelocityWindow::sendMsgEvt() {
    geometry_msgs::Twist msg;

    double linearX = this->editLinear->text().toDouble();
    double angularZ = this->editAngular->text().toDouble();

    msg.linear.x = linearX;
    msg.angular.z = angularZ * M_PI / 180;

    this->publisher.publish(msg);
}

File: 06ros/ws01/first_ws/src/ctrl_turtle_pkg/scripts/VelocityWindow.py
#! /usr/bin/env python
# coding:utf-8
from PyQt5.QtWidgets import (
    QWidget, QFormLayout, QLineEdit, QPushButton,
)
import rospy
from geometry_msgs.msg import Twist


class VelocityWindow(QWidget):
    def __init__(self, publisher):
        super(VelocityWindow, self).__init__()
        self.publisher = publisher
        self.setWindowTitle("小乌龟速度控制")

        self.layout = QFormLayout(self)
        self.linearXEdit = QLineEdit("0.0")
        self.angularZEdit = QLineEdit("0.0")
        self.layout.addRow("线速度", self.linearXEdit)
        self.layout.addRow("角速度", self.angularZEdit)

        self.sendBtn = QPushButton("发送")
        self.layout.addRow(self.sendBtn)

        # slot func
        self.sendBtn.clicked.connect(self.send_msg)

    def send_msg(self):
        if not isinstance(self.publisher, rospy.Publisher):
            return
        msg = Twist()
        linearX = float(self.linearXEdit.text())
        angularZ = float(self.angularZEdit.text())

        msg.linear.x = linearX
        msg.angular.z = angularZ * 3.14 / 180

        self.publisher.publish(msg)